Make download placeholder and final URL available to client

pvollan opened the pull request:
#### 5078adc67152c261c7fb6c83de4a4fc4540c8164
<pre>
Make download placeholder and final URL available to client
<a href="https://bugs.webkit.org/show_bug.cgi?id=278688">https://bugs.webkit.org/show_bug.cgi?id=278688</a>
<a href="https://rdar.apple.com/134745889">rdar://134745889</a>

Reviewed by Alex Christensen.

In some cases, it is important for the download client to know the location of the
placeholder file created for ongoing downloads as well as the final location. This
patch prepares for this by sending the placeholder URL and final URL from the
Networking process back to the client in the UI process. In order for the UI process
to be able to relocate the file if it is being moved at some point, we also send
bookmark data for the URL. Additionally, this patch adds a URL parameter to the
decidePlaceholderPolicy delegate completion handler, since a client should be able
to specify a custom URL if it chooses to manage the placeholder on its own.

* Source/WebKit/NetworkProcess/Downloads/Download.cpp:
(WebKit::Download::didFinish):
* Source/WebKit/NetworkProcess/Downloads/Download.h:
* Source/WebKit/NetworkProcess/Downloads/DownloadManager.cpp:
(WebKit::DownloadManager::publishDownloadProgress):
(WebKit::DownloadManager::clientReceivedPlaceholderURL): Deleted.
* Source/WebKit/NetworkProcess/Downloads/DownloadManager.h:
* Source/WebKit/NetworkProcess/Downloads/cocoa/DownloadCocoa.mm:
(WebKit::Download::publishProgress):
(WebKit::Download::setPlaceholderURL):
(WebKit::Download::setFinalURL):
(WebKit::Download::startUpdatingProgress const):
(WebKit::Download::clientReceivedPlaceholderURL): Deleted.
(WebKit::Download::startObservingProgress const): Deleted.
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::publishDownloadProgress):
(WebKit::NetworkProcess::clientReceivedPlaceholderURL): Deleted.
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/UIProcess/API/APIDownloadClient.h:
(API::DownloadClient::didReceivePlaceholderURL):
(API::DownloadClient::didReceiveFinalURL):
* Source/WebKit/UIProcess/API/Cocoa/WKDownload.mm:
* Source/WebKit/UIProcess/API/Cocoa/WKDownloadDelegatePrivate.h:
* Source/WebKit/UIProcess/Downloads/DownloadProxy.h:
* Source/WebKit/UIProcess/Downloads/DownloadProxy.messages.in:
* Source/WebKit/UIProcess/Downloads/DownloadProxyCocoa.mm:
(WebKit::DownloadProxy::didReceivePlaceholderURL):
(WebKit::DownloadProxy::didReceiveFinalURL):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Download.mm:
(-[ProgressObserver init]):
(-[ProgressObserver observeValueForKeyPath:ofObject:change:context:]):
(-[ProgressObserver waitForProgress]):
(tempFileThatDoesExist):
(TestWebKitAPI::PlaceholderPolicyEnable)):
(TestWebKitAPI::PlaceholderPolicyDisable)):
* Tools/TestWebKitAPI/cocoa/TestDownloadDelegate.h:
* Tools/TestWebKitAPI/cocoa/TestDownloadDelegate.mm:
(-[TestDownloadDelegate _download:didReceivePlaceholderURL:completionHandler:]):
(-[TestDownloadDelegate _download:didReceiveFinalURL:]):

Canonical link: <a href="https://commits.webkit.org/283386@main">https://commits.webkit.org/283386@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f0396d90433fe518800195a8196f0d513e2114f4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66146 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45519 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18765 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70178 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16756 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/68264 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53318 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17037 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/53085 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11667 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69213 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41990 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57266 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33723 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38661 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/14645 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15632 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/60555 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14990 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71880 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10101 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/14397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/60400 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10133 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57329 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60691 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8345 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1977 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10013 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/41327 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42403 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/43586 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42147 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/WKDownloadDelegatePrivate.h:
```diff
@@ -50,9 +50,25 @@ WK_SWIFT_UI_ACTOR
+ This is useful if the client maintains it's own placeholder file.
  */
-- (void)_download:(WKDownload *)download decidePlaceholderPolicy:(void (^)(_WKPlaceholderPolicy))completionHandler WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
+- (void)_download:(WKDownload *)download decidePlaceholderPolicy:(void (^)(_WKPlaceholderPolicy, NSURL *))completionHandler WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
```
Can we replace _WKPlaceholderPolicy with the nullness of the URL?

pvollan replied:
That's a good point! I don't believe we can, since we need to be able to distinguish between the use cases where the client does not want a placeholder at all, and the use case where the client is maintaining the placeholder and does not want the placeholder provided by the system.

Thanks for reviewing! I will address the comments in the upcoming patch.

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/WKDownloadDelegatePrivate.h:
```diff
@@ -50,9 +50,25 @@ WK_SWIFT_UI_ACTOR
+ @param completionHandler The completion handler that should be called by the client in response to this call. 
+ @discussion The placeholder URL will normally refer to a file in the Downloads directory
+ */
+- (void)_download:(WKDownload *)download didReceivePlaceholderURL:(NSURL *)url completionHandler:(void (^)(void))completionHandler WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
```
Are these callbacks just informative, or do we actually need to wait for something to happen before continuing?

pvollan replied:
Good point! The callback in didReceivePlaceholderURL  is used to send a message back to the Network process to inform that the client has received the placeholder, but the callback for didReceiveFinalURL is just informative, and can be removed.

achristensen07 commented on Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
```diff
@@ -55,6 +55,7 @@ messages -> NetworkProcess LegacyReceiver {
 #if PLATFORM(COCOA)
 #if HAVE(MODERN_DOWNLOADPROGRESS)
     PublishDownloadProgress(WebKit::DownloadID downloadID, URL url, std::span<const uint8_t> bookmarkData, enum:bool WebKit::UseDownloadPlaceholder useDownloadPlaceholder)
+    ClientReceivedPlaceholderURL(WebKit::DownloadID downloadID)
```
This should be the reply and sendWithAsyncReply instead of making this a separate message type.

achristensen07 commented on Tools/TestWebKitAPI/cocoa/TestDownloadDelegate.mm:
```diff
@@ -59,10 +59,10 @@ - (void)download:(WKDownload *)download didReceiveAuthenticationChallenge:(NSURL
 {
     _callbackRecord.append(DownloadCallback::DecidePlaceholderPolicy);
-    completionHandler(_WKPlaceholderPolicyDisable);
+    completionHandler(_WKPlaceholderPolicyDisable, nil);
```
The new delegate callbacks need to be tested.  We should also have a test with a non-null URL here and if possible we should verify that something happens with that URL.

achristensen07 commented on Source/WebKit/UIProcess/Downloads/DownloadProxyCocoa.mm:
```diff
@@ -54,4 +56,23 @@
+
+void DownloadProxy::didReceiveFinalURL(const URL& finalURL, std::span<const uint8_t> bookmarkData)
+{
+    m_client->didReceiveFinalURL(*this, finalURL, bookmarkData, [] {
```
This shouldn't have a completion handler.

achristensen07 commented on Source/WebKit/NetworkProcess/Downloads/cocoa/DownloadCocoa.mm:
```diff
@@ -122,6 +128,62 @@
+    if (usingSecurityScopedURL)
+        [placeholderURL stopAccessingSecurityScopedResource];
+
+    send(Messages::DownloadProxy::DidReceivePlaceholderURL(placeholderURL, span(data)));
```
sendWithAsyncReply.

achristensen07 commented on Source/WebKit/NetworkProcess/Downloads/cocoa/DownloadCocoa.mm:
```diff
@@ -122,6 +128,62 @@
+    BOOL usingSecurityScopedURL = [finalURL startAccessingSecurityScopedResource];
+
+    NSError *error = nil;
+    NSData *data = [finalURL bookmarkDataWithOptions:0 includingResourceValuesForKeys:nil relativeToURL:nil error:&error];
```
Does this need to be called in the network process and passed separately?

pvollan replied:
Yes, we send bookmark data for the URL, in case the client wants to be able to locate the file after it has been moved or deleted.

achristensen07 commented on Source/WebKit/NetworkProcess/Downloads/cocoa/DownloadCocoa.mm:
```diff
@@ -122,6 +128,62 @@
     }
 }
+
+void Download::setPlaceholderURL(NSURL *placeholderURL)
```
Nothing calls this function.  It seems like it should be asynchronous and something should be waiting for it before continuing.

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/WKDownload.mm:
```diff
@@ -202,6 +204,50 @@ void processDidCrash(WebKit::DownloadProxy& download) final
+
+        NSURL *placeholderURL = urlFromBookmark ? urlFromBookmark.get() : (NSURL *)url;
+
+        [m_delegate _download:wrapper(download) didReceivePlaceholderURL:placeholderURL completionHandler:makeBlockPtr([completionHandler = WTFMove(completionHandler)] () mutable {
```
You might be able to just call makeBlockPtr(WTFMove(completionHandler)).get()

achristensen07 commented on Source/WebKit/NetworkProcess/Downloads/Download.h:
```diff
@@ -129,10 +136,12 @@ class Download : public IPC::MessageSender, public CanMakeWeakPtr<Download>, pub
     RetainPtr<NSData> m_bookmarkData;
     RetainPtr<NSURL> m_bookmarkURL;
+    bool m_hasReceivedPlaceholderURL { false };
+    bool m_hasReceivedFinalURL { false };
```
These boolean states seem like they shouldn't be necessary with proper use of sendWithAsyncReply, but I'm not sure.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/Download.mm:
```diff
@@ -2278,6 +2323,92 @@ HTTPServer server({
+            completionHandler(expectedDownloadFile);
+        };
+        delegate.get().decidePlaceholderPolicy = ^(WKDownload *, void (^completionHandler)(_WKPlaceholderPolicy, NSURL *)) {
+            completionHandler(_WKPlaceholderPolicyDisable, tempFileThatDoesExist());
```
Can we check the contents of this file at the end of the test to make sure it is as expected?

pvollan replied:
This file will actually never be written to. It's just a placeholder file that progress is being published against.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/Download.mm:
```diff
@@ -2278,6 +2323,92 @@ HTTPServer server({
+    [webView setNavigationDelegate:delegate.get()];
+
+    auto progressObserver = adoptNS([[ProgressObserver alloc] init]);
+    auto progress = adoptNS([[NSProgress alloc] init]);
```
Don't we want to have progressObserver observe download.progress instead of allocating a new NSProgress object and not hooking it up to anything?  I don't understand how waitForProgress ever returns.

pvollan replied:
I think we want to observe the progress of the placeholder, since this is where progress is published in this case. In the latest patch, I set the fileURL property to the placeholder URL, in order to hook this up.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/Download.mm:
```diff
@@ -2248,6 +2288,11 @@ HTTPServer server({
 #if HAVE(MODERN_DOWNLOADPROGRESS)
+static bool modernDownloadProgressEnabled()
+{
+    return CFPreferencesGetAppBooleanValue(CFSTR("EnableModernDownloadProgress"), CFSTR("com.apple.WebKit"), NULL);
```
I see this is used in the network process in a few places.  We generally don't like doing that in a sandboxed process, preferring to use NetworkProcessCreationParameters or NetworkSessionCreationParameters instead.  Also, we don't have test coverage for both on and off.  I would expect a CFPreferencesSetAppBooleanValue to see what happens with it on and off.

pvollan replied:
Ah, yes, that's a good point. I will send this setting from the UI process to the Networking process in the upcoming patch.

The runtime setting will be removed when we enable this feature. Should we still create test cases with it on and off?

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/Download.mm:
```diff
@@ -1267,6 +1297,16 @@ static void checkFileContents(NSURL *file, const String& expectedContents)
+{
+    NSURL *tempDir = [NSURL fileURLWithPath:[NSTemporaryDirectory() stringByAppendingPathComponent:@"DownloadTest"] isDirectory:YES];
+    [[NSFileManager defaultManager] createDirectoryAtURL:tempDir withIntermediateDirectories:YES attributes:nil error:nil];
+    NSURL *file = [tempDir URLByAppendingPathComponent:@"example.download"];
```
This makes the directory exist but not the file, if I'm reading it correctly.

pvollan replied:
Ah, good catch, will fix in the upcoming patch!

achristensen07 commented on Source/WebKit/UIProcess/Downloads/DownloadProxyCocoa.mm:
```diff
@@ -54,4 +56,18 @@
+#if HAVE(MODERN_DOWNLOADPROGRESS)
+void DownloadProxy::didReceivePlaceholderURL(const URL& placeholderURL, std::span<const uint8_t> bookmarkData, CompletionHandler<void()>&& completionHandler)
+{
+    m_client->didReceivePlaceholderURL(*this, placeholderURL, bookmarkData, [completionHandler = WTFMove(completionHandler)] mutable {
```
No need to wrap completionHandler in a lambda that just captures and calls it.

achristensen07 commented on Source/WebKit/UIProcess/Downloads/DownloadProxyCocoa.mm:
```diff
@@ -26,6 +26,8 @@
 #import "config.h"
 #import "DownloadProxy.h"
 
+#import "APIData.h"
```
This doesn't seem used.

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/WKDownload.mm:
```diff
@@ -148,20 +150,20 @@ void decideDestinationWithSuggestedFilename(WebKit::DownloadProxy& download, con
     {
         if (!m_respondsToDecidePlaceholderPolicy) {
-            completionHandler(WebKit::UseDownloadPlaceholder::No);
+            completionHandler(WebKit::UseDownloadPlaceholder::No, WTF::URL());
```
You can probably just use { } for places where we are just sending an empty thing.

achristensen07 commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -2168,10 +2168,17 @@ void NetworkProcess::findPendingDownloadLocation(NetworkDataTask& networkDataTas
+                publishURL = WTFMove(alternatePlaceholderURL);
+            else
+                publishURL = URL::fileURLWithFileSystemPath(destination);
+            publishDownloadProgress(downloadID, publishURL, std::span<const uint8_t>(), usePlaceholder);
```
{ } instead of std::span<const uint8_t>()

achristensen07 commented on Source/WebKit/NetworkProcess/Downloads/cocoa/DownloadCocoa.mm:
```diff
@@ -122,6 +134,59 @@
+{
+    auto startObservingSelector = @selector(startObserving);
+    if ([m_progress respondsToSelector:startObservingSelector])
+        [m_progress performSelector:startObservingSelector];
```
We usually just call the selector directly after a respondsToSelector check:

if ([m_progress respondsToSelector:selector(startObserving)])
        [m_progress startObserving];

achristensen07 commented on Source/WebKit/NetworkProcess/Downloads/Download.cpp:
```diff
@@ -141,6 +141,13 @@ void Download::didFinish()
+#if HAVE(MODERN_DOWNLOADPROGRESS)
+    // If we are using a placeholder, we will call didFinish in Download::setFinalURL.
+    // This is because m_downloadManager->downloadFinished will delete the Download object.
+    if (m_isUsingPlaceholder)
```
This split flow isn't great.  It looks like we don't revoke the sandbox extension in the setFinalURL flow.  Do we know that setFinalURL will always be called after didFinish if this is true, or could it be called before?

pvollan replied:
We will still revoke the sandbox extension, since this is done in Download::didFinish(), which we call from Download::setFinalURL. Download::setFinalURL should always be called after didFinish, since it is based on the progress we report.

achristensen07 commented on Source/WebKit/NetworkProcess/Downloads/cocoa/DownloadCocoa.mm:
```diff
@@ -122,6 +134,55 @@
+    BOOL usingSecurityScopedURL = [placeholderURL startAccessingSecurityScopedResource];
+
+    NSError *error = nil;
+    NSData *data = [placeholderURL bookmarkDataWithOptions:(1<<24) includingResourceValuesForKeys:nil relativeToURL:nil error:&error];
```
Can we give 1<<24 a name and maybe use an SPI header at some point?

pvollan replied:
This option is not required, and I have removed it in the latest patch.

achristensen07 commented on Source/WebKit/NetworkProcess/Downloads/cocoa/DownloadCocoa.mm:
```diff
@@ -78,15 +79,18 @@
+#if HAVE(MODERN_DOWNLOADPROGRESS)
+    m_bookmarkURL = nil;
+    [m_progress unpublish];
+    [m_progress cancel];
```
I think we probably shouldn't call cancel here unless it's actually needed for something.  It should've already been cancelled by this point.

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/WKDownload.mm:
```diff
@@ -202,14 +202,54 @@ void processDidCrash(WebKit::DownloadProxy& download) final
+        if (bookmarkResolvingError || bookmarkDataIsStale)
+            RELEASE_LOG_ERROR(Network, "Failed to resolve URL from bookmark data");
+
+        NSURL *finalURL = urlFromBookmark ? urlFromBookmark.get() : (NSURL *)url;
```
I think ObjC can use the ?: operator, and I think the explicit cast to (NSURL *) might not be necessary.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/Download.mm:
```diff
@@ -853,6 +853,36 @@ - (void)_download:(_WKDownload *)download didReceiveData:(uint64_t)length
+
+- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context
+{
+    receivedProgress = true;
```
Could we check that the key path is as expected and the progress is more than 0% done?

pvollan committed:
Make download placeholder and final URL available to client
https://bugs.webkit.org/show_bug.cgi?id=278688
rdar://134745889

Reviewed by Alex Christensen.

In some cases, it is important for the download client to know the location of the
placeholder file created for ongoing downloads as well as the final location. This
patch prepares for this by sending the placeholder URL and final URL from the
Networking process back to the client in the UI process. In order for the UI process
to be able to relocate the file if it is being moved at some point, we also send
bookmark data for the URL. Additionally, this patch adds a URL parameter to the
decidePlaceholderPolicy delegate completion handler, since a client should be able
to specify a custom URL if it chooses to manage the placeholder on its own.

* Source/WebKit/NetworkProcess/Downloads/Download.cpp:
(WebKit::Download::didFinish):
* Source/WebKit/NetworkProcess/Downloads/Download.h:
* Source/WebKit/NetworkProcess/Downloads/DownloadManager.cpp:
(WebKit::DownloadManager::publishDownloadProgress):
(WebKit::DownloadManager::clientReceivedPlaceholderURL): Deleted.
* Source/WebKit/NetworkProcess/Downloads/DownloadManager.h:
* Source/WebKit/NetworkProcess/Downloads/cocoa/DownloadCocoa.mm:
(WebKit::Download::publishProgress):
(WebKit::Download::setPlaceholderURL):
(WebKit::Download::setFinalURL):
(WebKit::Download::startUpdatingProgress const):
(WebKit::Download::clientReceivedPlaceholderURL): Deleted.
(WebKit::Download::startObservingProgress const): Deleted.
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::publishDownloadProgress):
(WebKit::NetworkProcess::clientReceivedPlaceholderURL): Deleted.
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/UIProcess/API/APIDownloadClient.h:
(API::DownloadClient::didReceivePlaceholderURL):
(API::DownloadClient::didReceiveFinalURL):
* Source/WebKit/UIProcess/API/Cocoa/WKDownload.mm:
* Source/WebKit/UIProcess/API/Cocoa/WKDownloadDelegatePrivate.h:
* Source/WebKit/UIProcess/Downloads/DownloadProxy.h:
* Source/WebKit/UIProcess/Downloads/DownloadProxy.messages.in:
* Source/WebKit/UIProcess/Downloads/DownloadProxyCocoa.mm:
(WebKit::DownloadProxy::didReceivePlaceholderURL):
(WebKit::DownloadProxy::didReceiveFinalURL):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Download.mm:
(-[ProgressObserver init]):
(-[ProgressObserver observeValueForKeyPath:ofObject:change:context:]):
(-[ProgressObserver waitForProgress]):
(tempFileThatDoesExist):
(TestWebKitAPI::PlaceholderPolicyEnable)):
(TestWebKitAPI::PlaceholderPolicyDisable)):
* Tools/TestWebKitAPI/cocoa/TestDownloadDelegate.h:
* Tools/TestWebKitAPI/cocoa/TestDownloadDelegate.mm:
(-[TestDownloadDelegate _download:didReceivePlaceholderURL:completionHandler:]):
(-[TestDownloadDelegate _download:didReceiveFinalURL:]):

Canonical link: https://commits.webkit.org/283386@main
